boards/nRF52840-DK/src/lib.rs

```diff
@@ -0,0 +1,580 @@
+//! Board support crate for the Nordic nRF52840-DK
+//!
+//! This crate is in early development. Not much to see here, right now.
+#![no_std]
+
+//#![deny(missing_docs)]
+#![deny(warnings)]
+
+pub use cortex_m;
+pub use cortex_m_rt;
+pub use embedded_hal;
+pub use nrf52840_hal as hal;
+
+/// Exports traits that are usually needed when using this crate
+pub mod prelude {
+    pub use nrf52840_hal::prelude::*;
+}
+
+// TODO:
+// pub mod debug;
+
+use nrf52840_hal::{
+    prelude::*,
+    gpio::{
+        p0::{
+            self,
+            P0_Pin
+        },
+        p1,
+        Floating,
+        Input,
+        Output,
+        PushPull,
+        PullUp,
+        Level,
+    },
+    nrf52840_pac::{
+        self as nrf52,
+        CorePeripherals,
+        Peripherals,
+    },
+    spim::{
+        self,
+        Spim
+    },
+    uarte::{
+        self,
+        Uarte,
+        Parity as UartParity,
+        Baudrate as UartBaudrate,
+    },
+};
+
+/// Provides access to all features of the nRF52840-DK board
+#[allow(non_snake_case)]
+#[allow(non_camel_case_types)]
+pub struct nRF52840DK {
```

Comment:
Just an idea: I've been thinking for a while that `struct`s such as this should be called `Board`, like the peripheral struct PAC crates (and at least one HAL crate) is called `Peripherals`. This would also remove the repetition in the fully qualified name.

Not sure about this. Any comments?

Reply:
I agree, this would make sense. I will change this up!

What I am not quite sure about yet too is how I make the pins stored in here available to different owners iside an application lateron. The Peripheral allows you to take certain pins from certain places afaik. This is not the case here, right?

Reply:
After calling `take` or `steal`, the application will own `nRF52840DK`/`Board`, which includes ownership of `Pins` within it. Owning the pins means they can be moved out or borrowed as required.

Does that answer your question? Not sure I understood it correctly.

Reply:
Yeah, but you cannot move the Pins out of Board as it would make the fields of the struct uninitialized which is forbidden ;) It would consume the entire struct as far as I understand it.

Reply:
That's not exactly right. Moving a field out of a struct does two things:
- It makes that field inaccessible.
- It will prevent any use that requires the whole struct.

What this means is, you can no longer move or borrow the whole struct (which including calling methods that require `self`/`&self`/`&mut self`). You can still access any other fields though. I'm also relatively sure that you can make the struct usable again as a whole by moving stuff back into the fields you previously moved out of.

The main guideline to pay attention to is, for a given struct, follow one style or the other. Either design your struct with open fields that can be moved out of, or add methods to it that require some form of `self`. Never mix both in the same struct, unless you explicitly want those methods to be uncallable while the struct is incomplete.

Reply:
Well, there can be issues with moving out.
For example

```
self.s = consume_and_return(self.s);
```

can be hard to get working. There is [take_mut](https://docs.rs/take_mut/0.2.2/take_mut/) to get this working tho.
I also made this working at some places, but I am not quite sure yet when it works and when not.
Surely I will make some experiments :)
But for starters I am trying to get the SPI flash working to test out the SPI :)

Reply:
Yep, worked absolutely great for me in the past =)

Reply:
Fixed.